In [2]:
import numpy as np
import pandas as pd
from sklearn.discriminant_analysis import LinearDiscriminantAnalysis, QuadraticDiscriminantAnalysis
from sklearn.model_selection import train_test_split
from sklearn.metrics import confusion_matrix, classification_report
from sklearn.preprocessing import StandardScaler

# Load Data

In [3]:
df = pd.read_csv("Question_2.csv")
df

accents         X1        X2        X3         X4        X5         X6  \
0        ES   7.071476 -6.512900  7.650800  11.150783 -7.657312  12.484021   
1        ES  10.982967 -5.157445  3.952060  11.529381 -7.638047  12.136098   
2        ES   7.827108 -5.477472  7.816257   9.187592 -7.172511  11.715299   
3        ES   6.744083 -5.688920  6.546789   9.000183 -6.924963  11.710766   
4        ES   5.836843 -5.326557  7.472265   8.847440 -6.773244  12.677218   
..      ...        ...       ...       ...        ...       ...        ...   
324      US  -0.525273 -3.868338  3.548304   1.496249  3.490753   5.849887   
325      US  -2.094001 -1.073113  1.217397  -0.550790  2.666547   7.449942   
326      US   2.116909 -4.441482  5.350392   3.675396  2.715876   3.682670   
327      US   0.299616  0.324844  3.299919   2.044040  3.634828   6.693840   
328      US   3.214254 -3.135152  1.122691   4.712444  5.926518   6.915566   

            X7         X8         X9        X10       X11       X12  
0   -11.709772   3.426596   1.462715  -2.812753  0.866538 -5.244274  
1   -12.036247   3.491943   0.595441  -4.508811  2.332147 -6.221857  
2   -13.847214   4.574075  -1.687559  -7.204041 -0.011847 -6.463144  
3   -12.374388   6.169879  -0.544747  -6.019237  1.358559 -6.356441  
4   -12.315061   4.416344   0.193500  -3.644812  2.151239 -6.816310  
..         ...        ...        ...        ...       ...       ...  
324  -7.747027   9.738836 -11.754543   7.129909  0.209947 -1.946914  
325  -6.418064  10.907098 -11.134323   6.728373  2.461446 -0.026113  
326  -4.500850  11.798565 -12.031005   7.566142 -0.606010 -2.245129  
327  -5.676224  12.000518 -11.912901   4.664406  1.197789 -2.230275  
328  -5.799727  10.858532 -11.659845  10.605734  0.349482 -5.983281  

[329 rows x 13 columns]

In [4]:
y = df['accents']
X = df.drop(columns='accents')

# Train test split and scale data

In [5]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

scaler = StandardScaler()
X_train = scaler.fit_transform(X_train)
X_test = scaler.fit_transform(X_test)


# Linear Discriminant Analysis

In [6]:
lda = LinearDiscriminantAnalysis()
lda.fit(X_train, y_train)
pred = lda.predict(X_test)

print("Confusion Matrix for Linear Discriminant Analysis")
print(confusion_matrix(y_test, pred))
print("\n\n========================")
print("Classification Report for Linear Discriminant Analysis \n")
print(classification_report(y_test, pred))

Confusion Matrix for Linear Discriminant Analysis
[[10  0  0  0  1  0]
 [ 0  2  0  2  0  4]
 [ 0  0  6  0  0  4]
 [ 0  0  1  4  1  6]
 [ 0  0  0  0 10  3]
 [ 1  3  1  0  1 39]]


Classification Report for Linear Discriminant Analysis 

              precision    recall  f1-score   support

          ES       0.91      0.91      0.91        11
          FR       0.40      0.25      0.31         8
          GE       0.75      0.60      0.67        10
          IT       0.67      0.33      0.44        12
          UK       0.77      0.77      0.77        13
          US       0.70      0.87      0.77        45

    accuracy                           0.72        99
   macro avg       0.70      0.62      0.64        99
weighted avg       0.71      0.72      0.70        99



# Quadratic Discriminant Analysis

In [7]:
qda = QuadraticDiscriminantAnalysis()
qda.fit(X_train, y_train)
pred = qda.predict(X_test)

print("Confusion Matrix for Quadratic Discriminant Analysis")
print(confusion_matrix(y_test, pred))
print("\n\n========================")
print("Classification Report for Quadratic Discriminant Analysis \n")
print(classification_report(y_test, pred))

Confusion Matrix for Quadratic Discriminant Analysis
[[ 8  0  0  0  0  3]
 [ 0  6  0  0  0  2]
 [ 0  0  4  0  0  6]
 [ 0  0  0  2  2  8]
 [ 0  1  0  0 10  2]
 [ 0  2  0  0  0 43]]


Classification Report for Quadratic Discriminant Analysis 

              precision    recall  f1-score   support

          ES       1.00      0.73      0.84        11
          FR       0.67      0.75      0.71         8
          GE       1.00      0.40      0.57        10
          IT       1.00      0.17      0.29        12
          UK       0.83      0.77      0.80        13
          US       0.67      0.96      0.79        45

    accuracy                           0.74        99
   macro avg       0.86      0.63      0.67        99
weighted avg       0.80      0.74      0.71        99



# Compare the performance of the classifiers in (b) and (c)

Based on the confusion matrix and the classification report, it can be seen that the accuracy of the Quadratic Discriminant Analysis (QDA) is better than the Linear Discriminant Analysis (LDA). Looking at the confusion matrix, LDA has more wrongly classified items as compared to QDA. If we look deeper into the classfication report, we can compare metrices such as precision, recall, f1, and accuracy. First of all, if we look at accuracy score which gives a broad idea on the model performance, here we can already compare the performance between the two models in that QDA has an accuracy of 0.74 which LDA is only at 0.72. 

If we look at the next metric which is precision. Precision is the ratio of $\frac{TP}{TP+FP}$, where TP is the true positive, and FP is the false positive. The precision is intuitively the ability of the classifier not to label as positive a sample that is negative. In that, a high precision model is able to reduce the false positives. Unlike accuracy which gives an overview of the model, precision score is given according to each class. In LDA, the highest precision score is 0.91 for class ES, and the lowest score is 0.4 for FR. As compared to QDA where for some classes such as ES, GE, and IT it is able to obtain a precision score of 1 and lowest precision is only at 0.67. QDA has been able to perfectly discriminate these classes and avoid unnecessary FPs. As compared to LDA which still struggles to discriminate between classes and picks up false positives.

Recall is another metric available in the classification report. Recall is the ratio of $\frac{TP}{TP+FN}$, where FN is false negative. The recall is intuitively the ability of the classifier to find all the positive samples, in a sense that high recall value indicates that the model is able to reduce false negatives and find large numbers of correctly sorted positives. For LDA, the recall value ranges 0.25 to 0.91, as compared to QDA which ranges 0.17 to 0.96. QDA has a wider spread of recall value compared to LDA meaning that it has higher FNs in its classifier.

Looking at the F1 score, which can be interpreted as a harmonic mean of the precision and recall. Often, in cases where the metric of choice is not reliant on either precision or recall, F1 score can be considered. Unlike certain cases such as medical diagnostic where higher recall is preferrable due to intention of minimizing false negative to avoid underdiagnosis. In this case F1 score can give a balanced trade off between the performance of the models. In LDA, the spread of the F1 score ranges 0.44 to 0.91, while QDA is 0.29 to 0.84. It can be seen that LDA performs better in general use as compared to the QDA eventhough the accuracy is lower.

In conclusion, it can be seen that when looking into specific performance metrics to analyze the performance of classifiers, it is important to take into account the use case. Be it; to prioritize identifying correctly labeled positive classes or to find all positive classes at the cost of false negatives. 